AsyncGuidance.md

```diff
@@ -738,6 +738,8 @@ public class PersonController : Controller
 }
 ```
 
+:bulb: **NOTE:** Using `ValueTask<T>` removes an allocation as most times we'd return the value from the cache.
```

Comment:
It doesn't actually. Using ValueTask only removes the allocation if:
- You implement a custom IValueTaskSource
- You enable to ValueTask pooling
- You return synchronously

Reply:
In this method we mostly await a *completed* cached task, which means we return synchronously. The compiler emits this code:

```cs
var awaiter = _cache.GetOrAdd(id, (key) => new AsyncLazy<Person>(() => _db.People.FindAsync(key))).Value.GetAwaiter();
if (awaiter.IsCompleted)
{
   builder.SetResult(awaiter.GetResult());
}
```

In this path, there's no call to `AsyncValueTaskMethodBuilder<TResult>.AwaitOnCompleted` which allocates the `Task`.

```cs
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System.Threading.Tasks;

[MemoryDiagnoser]
public class Benchmarks
{
    private static readonly Task<string> _cachedTask = Task.FromResult("s");

    [Benchmark(OperationsPerInvoke = 100)]
    public async ValueTask<string> ValueTaskMethod() => await _cachedTask;

    [Benchmark(OperationsPerInvoke = 100)]
    public async Task<string> TaskMethod() => await _cachedTask;

    public static void Main() => BenchmarkRunner.Run<Benchmarks>();
}
```

Results in:
```
|          Method |      Mean |     Error |    StdDev |  Gen 0 | Allocated |
|---------------- |----------:|----------:|----------:|-------:|----------:|
| ValueTaskMethod | 0.2603 ns | 0.0104 ns | 0.0267 ns |      - |         - |
|      TaskMethod | 0.2213 ns | 0.0051 ns | 0.0114 ns | 0.0002 |       1 B |
```

Reply:
It's absolutely not worth it 😄. I would say change the caching function to be ValueTask based (but that has subtle problems because we want to support multiple awaiters). Anything else is too much work for too little benefit.

Reply:
Can you elaborate why you think it's not worth it? IMO this is a good general guidance for async caching services.

How is it different than this recommendation:
> :white_check_mark: **GOOD** This example uses a `ValueTask<int>` to return the trivially computed value. It does not use any extra threads as a result. It also does not allocate an object on the managed heap.

Reply:
Is it different because we are using a `ConcurrentDictionary` in this case? and we need multiple awaiters for this?